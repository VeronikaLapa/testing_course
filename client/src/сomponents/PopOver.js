import React, {useEffect} from 'react';
import {makeStyles} from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import IconButton from '@material-ui/core/IconButton';
import MenuIcon from '@material-ui/icons/Menu';
import AccountCircle from '@material-ui/icons/AccountCircle';
import MenuItem from '@material-ui/core/MenuItem';
import Menu from '@material-ui/core/Menu';
import {Link} from "react-router-dom";
import {connect} from "react-redux";
import fetchUser from "../store/actions/fetchUser";
import getAuth from "../store/actions/getAuth";

const useStyles = makeStyles((theme) => ({
    root: {
        flexGrow: 1,
    },
    menuButton: {
        marginRight: theme.spacing(2),
    },
    title: {
        flexGrow: 1,
    },
}));

function PopOver({auth, name, getAuthUserAction}) {
    const classes = useStyles();
    const [anchorEl, setAnchorEl] = React.useState(null);
    const [open, setOpen] = React.useState(false);

    useEffect(() => {
        if (anchorEl && auth) {
            setOpen(true);
        } else {
            setOpen(false);
        }
        if (!auth) {
            setAnchorEl(null);
        }
    });

    const handleMenu = (event) => {
        console.log('handle menu');
        setAnchorEl(event.currentTarget);
        if (name == null) {
            //getAuthUserAction();
        }
        setOpen(true);

    };

    const handleClose = () => {
        setOpen(false);
        setAnchorEl(null);
    };


    return (
        <div>
            {auth && (
                <div>
                    <IconButton
                        aria-label="account of current user"
                        aria-controls="menu-appbar"
                        aria-haspopup="true"
                        onClick={handleMenu}
                        color="inherit"
                        id='avatar'
                    >
                        <AccountCircle/>
                    </IconButton>
                    <Menu
                        id="menu-appbar"
                        anchorEl={anchorEl}
                        anchorOrigin={{
                            vertical: 'top',
                            horizontal: 'right',
                        }}
                        keepMounted
                        transformOrigin={{
                            vertical: 'top',
                            horizontal: 'right',
                        }}
                        open={open}
                        onClose={handleClose}
                    >
                        <MenuItem onClick={handleClose}>{name}</MenuItem>
                        <MenuItem><Link to="/logout"><Typography color="error">Exit</Typography></Link></MenuItem>
                    </Menu>
                </div>
            )}
        </div>
    );
}

const mapStateToProps = store => {
    console.log(store); // посмотрим, что же у нас в store?
    return {
        auth: store.user.auth,
        name: store.user.user.name
    }
};
const mapDispatchToProps = dispatch => {
    return {
        getAuthUserAction: () => dispatch(getAuth()), // [1]
    }
};


export default connect(mapStateToProps, mapDispatchToProps)(PopOver);
