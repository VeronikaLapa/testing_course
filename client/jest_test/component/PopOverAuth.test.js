import configureStore from 'redux-mock-store'
import React from "react";
import {Provider} from "react-redux";
import {BrowserRouter, HashRouter, MemoryRouter, Router} from "react-router-dom";
import PopOver from "../../src/сomponents/PopOver";
import { shallow, mount } from 'enzyme';
import "../setupTest"
import {IconButton} from "@material-ui/core";
import Menu from "@material-ui/core/Menu";


describe('PopOver render with auth',()=> {
    const initialState = {user: {
            auth: true,
            message: '',
            pending: false,
            justLogin: false,
            token: '',
            user: {name: 'Ann'}}};
    const mockStore = configureStore();
    let store, container;
    let wrapper;
    let reduxState;

    beforeEach(() => {
        store = mockStore(initialState);
        reduxState = initialState;
    });

    it('check popover rendered', function () {
        //const wrapper = mount(<Provider store={store}><MemoryRouter><PopOver /></MemoryRouter></Provider>);
        const wrapper = mount(<Provider store={store}><MemoryRouter><PopOver /></MemoryRouter></Provider>);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(IconButton).length).toEqual(1);
        expect(wrapper.find(Menu).length).toEqual(1)
    });

    it('check menu open property on icon click', function () {
        const wrapper = mount(<Provider store={store}><MemoryRouter><PopOver /></MemoryRouter></Provider>);
        const popover = wrapper.find(PopOver);

        const menu = wrapper.find(Menu);
        const icon = wrapper.find(IconButton);

        expect(wrapper.find(Menu).prop("open")).toEqual(false);
        icon.simulate("click");
        expect(wrapper.find(Menu).prop("open")).toEqual(true);
    });

    it('close() close menu', function () {
        const wrapper = mount(<Provider store={store}><MemoryRouter><PopOver /></MemoryRouter></Provider>);

    })
});
