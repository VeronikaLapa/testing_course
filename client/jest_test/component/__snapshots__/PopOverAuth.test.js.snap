// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopOver render with auth check popover rendered 1`] = `ReactWrapper {}`;
